compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3318,7 +3318,8 @@ public void visit(BLangErrorConstructorExpr errorConstructorExpr, AnalyzerData d
         if (Types.getReferredType(detailType).tag == TypeTags.MAP) {
             BType errorDetailTypeConstraint = ((BMapType) Types.getReferredType(detailType)).constraint;
             for (BLangNamedArgsExpression namedArgExpr: namedArgs) {
-                if (!types.isAssignable(namedArgExpr.expr.getBType(), errorDetailTypeConstraint)) {
+                if (errorDetailTypeConstraint.tag == TypeTags.UNION &&
+                        !types.isAssignable(namedArgExpr.expr.getBType(), errorDetailTypeConstraint)) {
```

Comment:
In here `errorDetailTypeConstraint` is a union; its wrapped inside a type reference
```ballerina
type StringType string|int;

type ErrorR1 error<map<StringType>>;

function dummyTest() {
    ErrorR1 er1 = error ErrorR1("Custom error", errorCode = "104", application = {});
}
```